### PR TITLE
Harpy Flight System

### PR DESCRIPTION
**Pull request by gluesniffler**
# Description

This PR adds a generic system which gives an entity the ability to fly. Optionally increasing their speed in exchange for a continuous stamina drain, which can, and **will** stamcrit them if left unchecked. 

---

# Technical Details?

We normally dont have this section but I'd like to outline the changes since I messed with quite a few systems:

- Introduces a `FlightComponent` which can be added to any entity in YML, needs to be tied to an action with an event of type `ToggleFlightEvent` This component holds properties for:
    - Toggling animations on and off, either at the entity level or the layer level.
    - Altering shader animation properties
    - Altering speed, stamina drain, sounds played, delay between sounds, etc etc.
- Adds a `FlyingVisualizerSystem` that can take a given `AnimationKey` which points to a shader, and optionally can apply it to either the entire sprite, or a given layer. 
- Adds a check in `SharedGravitySystem` for making the entity weightless when it has the `FlightComponent` and is flying.
- Adds a check in `SharedCuffableSystem` to disable cuffing when the target has the `FlightComponent` and is flying.
- Introduces a new field in the `StaminaComponent` which serves as a dictionary for persistent drains, with the key being the source (UID) of where it came from. The drains can also indicate if they should apply the stamina slowdown or not (relevant for both this PR, and for an eventual sprinting PR)

---

<details><summary><h1>Media</h1></summary>
<p>

[Flight Demo](https://youtu.be/Wndv9hYaZ_s "Flight Demo")
</p>
</details>

---

# Changelog

:cl: Mocho
- add: Harpies are now able to fly on station for limited periods of time, moving faster at the cost of stamina.

### CODE REVIEW

File: Content.Client/Flight/Components/FlightVisualsComponent.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    /// </summary>
+    public bool AnimateLayer = false;
+    public int? TargetLayer;
+
+    [ViewVariables(VVAccess.ReadWrite)]
+    [DataField]
+    public string AnimationKey = "default";
+
+    [ViewVariables(VVAccess.ReadWrite)]
+    public ShaderInstance Shader = default!;
```
Comment (VMSolidus) [inline Content.Client/Flight/Components/FlightVisualsComponent.cs:41]:
```suggestion
    /// <summary>
    ///     How long does the animation last
    /// </summary>
    [DataField]
    public float Speed;

    /// <summary>
    ///     How far it goes in any direction.
    /// </summary>
    [DataField]
    public float Multiplier;

    /// <summary>
    ///     How much the limbs (if there are any) rotate.
    /// </summary>
    [DataField]
    public float Offset;

    /// <summary>
    ///     Are we animating layers or the entire sprite?
    /// </summary>
    public bool AnimateLayer = false;
    public int? TargetLayer;

    [DataField]
    public string AnimationKey = "default";

    [ViewVariables(VVAccess.ReadWrite)]
    public ShaderInstance Shader = default!;
```

File: Content.Client/Flight/FlightSystem.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        private void OnFlight(FlightEvent args)
+        {
+            Logger.Debug("Starting onFlight!");
+            var uid = GetEntity(args.Uid);
+            if (!_entityManager.TryGetComponent(uid, out SpriteComponent? sprite)
+            || !args.IsAnimated
+            || !_entityManager.TryGetComponent(uid, out FlightComponent? flight))
+                return;
```
Comment (VMSolidus) [inline Content.Client/Flight/FlightSystem.cs:26]:
```suggestion
            if (!_entityManager.TryGetComponent(uid, out SpriteComponent? sprite)
                || !args.IsAnimated
                || !_entityManager.TryGetComponent(uid, out FlightComponent? flight))
                return;
```

File: Content.Client/Flight/FlightSystem.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
+using Robust.Client.GameObjects;
+using Content.Shared.Flight;
+using Content.Shared.Flight.Events;
+using Content.Client.Flight.Components;
+
+namespace Content.Client.Flight
+{
```
Comment (VMSolidus) [inline Content.Client/Flight/FlightSystem.cs:7]:
```suggestion
namespace Content.Client.Flight;
```
You are *also* going to need to move the entire document down an indent after doing this. Ctrl-A and Shift-Tab to do the entire document all at once.

File: Content.Client/Flight/FlightSystem.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                {
+                    return index;
+                }
+                index++;
+            }
+
+            return null;
+        }
+    }
+}
```
Comment (VMSolidus) [inline Content.Client/Flight/FlightSystem.cs:75]:
```suggestion
```

File: Content.Client/Flight/FlightSystem.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                return null;
+
+            int index = 0;
+            foreach (var layer in sprite.AllLayers)
+            {
+                // This feels like absolute shitcode, isn't there a better way to check for it?
+                if (layer.Rsi?.Path.ToString() == targetLayer)
+                {
+                    return index;
+                }
```
Comment (VMSolidus) [inline Content.Client/Flight/FlightSystem.cs:68]:
```suggestion
                if (layer.Rsi?.Path.ToString() == targetLayer)
                    return index;
```

File: Content.Client/Flight/FlightSystem.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                    Offset = flight.ShaderOffset,
+                    Speed = flight.ShaderSpeed,
+                    TargetLayer = targetLayer,
+                };
+                AddComp(uid, comp);
+            }
+            if (!args.IsFlying)
+            {
+                RemComp<FlightVisualsComponent>(uid);
+            }
```
Comment (VMSolidus) [inline Content.Client/Flight/FlightSystem.cs:53]:
```suggestion
            if (!args.IsFlying)
                RemComp<FlightVisualsComponent>(uid);
```

File: Content.Client/Flight/FlyingVisualizerSystem.cs [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        if (!animateLayer)
+        {
+            entity.Comp.PostShader = shader;
+        }
+
+        if (animateLayer && layer is not null)
+        {
+            entity.Comp.LayerSetShader(layer.Value, shader);
+        }
+        entity.Comp.GetScreenTexture = shader is not null;
```
Comment (VMSolidus) [inline Content.Client/Flight/FlyingVisualizerSystem.cs:49]:
```suggestion
        if (animateLayer && layer is not null)
            entity.Comp.LayerSetShader(layer.Value, shader);

        entity.Comp.GetScreenTexture = shader is not null;
```

File: Content.Client/Flight/FlyingVisualizerSystem.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        entity.Comp.RaiseShaderEvent = shader is not null;
+    }
+
+    private void OnBeforeShaderPost(EntityUid uid, FlightVisualsComponent comp, ref BeforePostShaderRenderEvent args)
+    {
+        // This function can be used to modify the shader's values while its running.
+        SetValues(comp, comp.Speed, comp.Offset, comp.Multiplier);
+    }
```
Comment (VMSolidus) [inline Content.Client/Flight/FlyingVisualizerSystem.cs:57]:
```suggestion

    /// <summary>
    ///     This function can be used to modify the shader's values while its running.
    /// </summary>
    private void OnBeforeShaderPost(EntityUid uid, FlightVisualsComponent comp, ref BeforePostShaderRenderEvent args)
    {
        SetValues(comp, comp.Speed, comp.Offset, comp.Multiplier);
    }
```

File: Content.Server/Flight/FlightSystem.cs [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+using Content.Shared.Flight;
+using Content.Shared.Flight.Events;
+using Content.Shared.Mobs;
+using Content.Shared.Popups;
+using Content.Shared.Stunnable;
+using Content.Shared.Zombies;
+using Robust.Shared.Audio.Systems;
+
+namespace Content.Server.Flight
+{
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:14]:
```suggestion
namespace Content.Server.Flight;
```
Just like before, you're going to need to drop the entire file down a tab indent.

File: Content.Server/Flight/FlightSystem.cs [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+                ToggleActive(uid, false, component);
+                if (TryComp<StaminaComponent>(uid, out var stamina))
+                    Dirty(uid, stamina);
+            }
+        }
+
+        #endregion
+
+    }
+}
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:151]:
```suggestion
```

File: Content.Server/Flight/FlightSystem.cs [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+                if (!component.On)
+                    continue;
+
+                component.TimeUntilFlap -= frameTime;
+
+                if (component.TimeUntilFlap <= 0f)
+                {
+                    _audio.PlayPvs(component.FlapSound, uid);
+                    component.TimeUntilFlap = component.FlapInterval;
+                }
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:49]:
```suggestion
                if (component.TimeUntilFlap > 0f)
                    continue;
                    
                _audio.PlayPvs(component.FlapSound, uid);
                component.TimeUntilFlap = component.FlapInterval;
```

File: Content.Server/Flight/FlightSystem.cs [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+                        BreakOnUserMove = true,
+                        BreakOnDamage = true,
+                        NeedHand = true
+                    };
+
+                    if (!_doAfter.TryStartDoAfter(doAfterArgs))
+                    {
+                        return;
+                    }
+                }
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:74]:
```suggestion
                if (!CanFly(uid, component))
                    return;
                    
                var doAfterArgs = new DoAfterArgs(EntityManager, 
                    uid, component.ActivationDelay, 
                    new FlightDoAfterEvent(), uid, target: uid)
                {
                    BlockDuplicate = true,
                    BreakOnTargetMove = true,
                    BreakOnUserMove = true,
                    BreakOnDamage = true,
                    NeedHand = true
                };

                if (!_doAfter.TryStartDoAfter(doAfterArgs))
                    return;
```

File: Content.Server/Flight/FlightSystem.cs [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+            }
+            else
+            {
+                ToggleActive(uid, false, component);
+            }
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:79]:
```suggestion
            else
                ToggleActive(uid, false, component);
```

File: Content.Server/Flight/FlightSystem.cs [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+
+            // Tell the user that they can not fly.
+            if (cuffed || zombified)
+            {
+                _popupSystem.PopupEntity(Loc.GetString("no-flight-while-restrained"), uid, uid, PopupType.Medium);
+                return false;
+            }
+
+            return true;
+        }
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:108]:
```suggestion
        private bool CanFly(EntityUid uid, FlightComponent component)
        {
            if (TryComp<CuffableComponent>(uid, out var cuffableComp) && !cuffableComp.CanStillInteract)
            {
                _popupSystem.PopupEntity(Loc.GetString("no-flight-while-restrained"), uid, uid, PopupType.Medium);
                return false;
            }
            
            if (HasComp<ZombieComponent>(uid))
            {
                _popupSystem.PopupEntity(Loc.GetString("no-flight-while-zombified"), uid, uid, PopupType.Medium);
                return false;
            }
            return true;
        }
```
You're also going to need to add a string for no-flight-while-zombified.

File: Content.Server/Flight/FlightSystem.cs [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+        private void OnMobStateChangedEvent(EntityUid uid, FlightComponent component, MobStateChangedEvent args)
+        {
+            if (component.On && args.NewMobState is MobState.Critical or MobState.Dead)
+                ToggleActive(args.Target, false, component);
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:113]:
```suggestion
            if (!component.On 
                || args.NewMobState is MobState.Critical or MobState.Dead)
                return;
                
            ToggleActive(args.Target, false, component);
```

File: Content.Server/Flight/FlightSystem.cs [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        }
+
+        private void OnZombified(EntityUid uid, FlightComponent component, ref EntityZombifiedEvent args)
+        {
+            if (component.On)
+            {
+                ToggleActive(args.Target, false, component);
+                if (TryComp<StaminaComponent>(uid, out var stamina))
+                    Dirty(uid, stamina);
+            }
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:123]:
```suggestion
            if (!component.On)
                return;

            ToggleActive(args.Target, false, component);
            if (!TryComp<StaminaComponent>(uid, out var stamina))
                return;
                
            Dirty(uid, stamina);
```

File: Content.Server/Flight/FlightSystem.cs [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+        }
+
+        private void OnKnockedDown(EntityUid uid, FlightComponent component, ref KnockedDownEvent args)
+        {
+            if (component.On)
+                ToggleActive(uid, false, component);
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:129]:
```suggestion
            if (!component.On)
                return;

            ToggleActive(uid, false, component);
```

File: Content.Server/Flight/FlightSystem.cs [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+        }
+
+        private void OnStunned(EntityUid uid, FlightComponent component, ref StunnedEvent args)
+        {
+            if (component.On)
+                ToggleActive(uid, false, component);
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:135]:
```suggestion
            if (!component.On)
                return;
                
            ToggleActive(uid, false, component);
```

File: Content.Server/Flight/FlightSystem.cs [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+
+        private void OnSleep(EntityUid uid, FlightComponent component, ref SleepStateChangedEvent args)
+        {
+            if (args.FellAsleep && component.On)
+            {
+                ToggleActive(uid, false, component);
+                if (TryComp<StaminaComponent>(uid, out var stamina))
+                    Dirty(uid, stamina);
+            }
+        }
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:146]:
```suggestion
        {
            if (!component.On 
                || !args.FellAsleep)
                return;

            ToggleActive(uid, false, component);
            if (!TryComp<StaminaComponent>(uid, out var stamina))
                return;
                
            Dirty(uid, stamina);
        }
```

File: Content.Server/Flight/FlightSystem.cs [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+
+        #endregion
+
```
Comment (VMSolidus) [inline Content.Server/Flight/FlightSystem.cs:149]:
```suggestion
        #endregion
```

File: Content.Shared/Damage/Components/StaminaComponent.cs [hunk 21]
```diff
@@ -39,6 +39,13 @@ public sealed partial class StaminaComponent : Component
     [ViewVariables(VVAccess.ReadWrite), DataField, AutoNetworkedField]
     public float CritThreshold = 100f;
 
+    /// <summary>
+    /// A dictionary of active stamina drains, with the key being the source of the drain,
+    /// DrainRate how much it changes per tick, and modifiesSpeed if it should slow down the user.
+    /// </summary>
+    [DataField("activeDrains"), AutoNetworkedField]
+    public Dictionary<EntityUid, (float DrainRate, bool ModifiesSpeed)> ActiveDrains = new();
```
Comment (VMSolidus) [inline Content.Shared/Damage/Components/StaminaComponent.cs:47]:
```suggestion
    /// <summary>
    ///     A dictionary of active stamina drains, with the key being the source of the drain,
    ///     DrainRate how much it changes per tick, and modifiesSpeed if it should slow down the user.
    /// </summary>
    [DataField, AutoNetworkedField]
    public Dictionary<EntityUid, (float DrainRate, bool ModifiesSpeed)> ActiveDrains = new();
```

File: Content.Shared/Damage/Systems/StaminaSystem.cs [hunk 22]
```diff
@@ -328,27 +328,52 @@ public void TakeStaminaDamage(EntityUid uid, float value, StaminaComponent? comp
...
+
+        if (enabled)
+        {
+            stamina.ActiveDrains[actualSource] = (drainRate, modifiesSpeed);
+            EnsureComp<ActiveStaminaComponent>(target);
+        }
+        else
+        {
+            stamina.ActiveDrains.Remove(actualSource);
+        }
```
Comment (VMSolidus) [inline Content.Shared/Damage/Systems/StaminaSystem.cs:347]:
```suggestion
        else
            stamina.ActiveDrains.Remove(actualSource);
```

**Comment by v0idRift** [comment]
Love this! Excited to see how this enhances gameplay! Birb

File: Content.Shared/Damage/Systems/StaminaSystem.cs [hunk 23]
```diff
@@ -328,27 +328,52 @@ public void TakeStaminaDamage(EntityUid uid, float value, StaminaComponent? comp
...
                 continue;
             }
-
+            if (comp.ActiveDrains.Count > 0)
+            {
+                foreach (var (source, (drainRate, modifiesSpeed)) in comp.ActiveDrains)
+                {
+                    TakeStaminaDamage(uid, drainRate * frameTime, comp, source: source, visual: false, allowsSlowdown: modifiesSpeed);
+                }
+            }
```
Comment (VMSolidus) [inline Content.Shared/Damage/Systems/StaminaSystem.cs:376]:
```suggestion
            if (comp.ActiveDrains.Count > 0)
                foreach (var (source, (drainRate, modifiesSpeed)) in comp.ActiveDrains)
                    TakeStaminaDamage(uid, 
                        drainRate * frameTime, 
                        comp, 
                        source: source, 
                        visual: false, 
                        allowsSlowdown: modifiesSpeed);
```

File: Content.Shared/Damage/Systems/StaminaSystem.cs [hunk 24]
```diff
@@ -363,8 +388,12 @@ public override void Update(float frameTime)
             }
 
             comp.NextUpdate += TimeSpan.FromSeconds(1f);
-            TakeStaminaDamage(uid, -comp.Decay, comp);
-            Dirty(comp);
+            // If theres no active drains, recover stamina.
+            if (comp.ActiveDrains.Count == 0)
+            {
+                TakeStaminaDamage(uid, -comp.Decay, comp);
+            }
```
Comment (VMSolidus) [inline Content.Shared/Damage/Systems/StaminaSystem.cs:395]:
```suggestion
            if (comp.ActiveDrains.Count == 0)
                TakeStaminaDamage(uid, -comp.Decay, comp);

```

File: Content.Shared/Flight/Events.cs [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+        public bool IsAnimated { get; }
+        public FlightEvent(NetEntity uid, bool isFlying, bool isAnimated)
+        {
+            Uid = uid;
+            IsFlying = isFlying;
+            IsAnimated = isAnimated;
+        }
+    }
+
+}
```
Comment (VMSolidus) [inline Content.Shared/Flight/Events.cs:30]:
```suggestion
namespace Content.Shared.Flight.Events;

[Serializable, NetSerializable]
public sealed partial class DashDoAfterEvent : SimpleDoAfterEvent { }

[Serializable, NetSerializable]
public sealed partial class FlightDoAfterEvent : SimpleDoAfterEvent { }

[Serializable, NetSerializable]
public sealed class FlightEvent : EntityEventArgs
{
    public NetEntity Uid { get; }
    public bool IsFlying { get; }
    public bool IsAnimated { get; }
    public FlightEvent(NetEntity uid, bool isFlying, bool isAnimated)
    {
        Uid = uid;
        IsFlying = isFlying;
        IsAnimated = isAnimated;
    }
}
```

File: Content.Shared/Flight/FlightComponent.cs [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+
+        /// <summary>
+        /// Time between sounds being played
+        /// </summary>
+        [DataField("flapInterval")]
+        public float FlapInterval = 1.0f;
+
+        public float TimeUntilFlap;
+    }
+}
```
Comment (VMSolidus) [inline Content.Shared/Flight/FlightComponent.cs:110]:
```suggestion
namespace Content.Shared.Flight;

/// <summary>
///     Adds an action that allows the user to become temporarily
///     weightless at the cost of stamina and hand usage.
/// </summary>
[RegisterComponent, NetworkedComponent(), AutoGenerateComponentState]
public sealed partial class FlightComponent : Component
{
    [DataField(customTypeSerializer: typeof(PrototypeIdSerializer<EntityPrototype>))]
    public string? ToggleAction = "ActionToggleFlight";

    [DataField, AutoNetworkedField]
    public EntityUid? ToggleActionEntity;

    /// <summary>
    ///     Is the user flying right now?
    /// </summary>
    [DataField, AutoNetworkedField]
    public bool On;

    /// <summary>
    ///     Stamina drain per second when flying
    /// </summary>
    [DataField, AutoNetworkedField]
    public float StaminaDrainRate = 6.0f;

    /// <summary>
    ///     DoAfter delay until the user becomes weightless.
    /// </summary>
    [DataField, AutoNetworkedField]
    public float ActivationDelay = 1.0f;

    /// <summary>
    ///     Speed modifier while in flight
    /// </summary>
    [DataField, AutoNetworkedField]
    public float SpeedModifier = 2.0f;

    /// <summary>
    ///     Path to a sound specifier or collection for the noises made during flight
    /// </summary>
    [DataField]
    public SoundSpecifier FlapSound = new SoundCollectionSpecifier("WingFlaps");

    /// <summary>
    ///     Is the flight animated?
    /// </summary>
    [DataField]
    public bool IsAnimated = true;

    /// <summary>
    ///     Does the animation animate a layer?.
    /// </summary>
    [DataField]
    public bool IsLayerAnimated;

    /// <summary>
    ///     Which RSI layer path does this animate?
    /// </summary>
    [DataField]
    public string? Layer;

    /// <summary>
    ///     Whats the speed of the shader?
    /// </summary>
    [DataField]
    public float ShaderSpeed = 6.0f;

    /// <summary>
    ///     How much are the values in the shader's calculations multiplied by?
    /// </summary>
    [DataField]
    public float ShaderMultiplier = 0.01f;

    /// <summary>
    ///     What is the offset on the shader?
    /// </summary>
    [DataField]
    public float ShaderOffset = 0.25f;

    /// <summary>
    ///     What animation does the flight use?
    /// </summary>

    [DataField]
    public string AnimationKey = "default";

    /// <summary>
    ///     Time between sounds being played
    /// </summary>
    [DataField]
    public float FlapInterval = 1.0f;

    public float TimeUntilFlap;
}
```

File: Content.Shared/Flight/SharedFlightSystem.cs [hunk 27]
```diff
@@ -0,0 +1,110 @@
+
```
Comment (VMSolidus) [inline Content.Shared/Flight/SharedFlightSystem.cs:1]:
```suggestion
```

File: Content.Shared/Flight/SharedFlightSystem.cs [hunk 28]
```diff
@@ -0,0 +1,110 @@
...
+            args.ModifySpeed(component.SpeedModifier, component.SpeedModifier);
+        }
+
+        #endregion
+    }
+    public sealed partial class ToggleFlightEvent : InstantActionEvent
+    {
+    }
+}
```
Comment (VMSolidus) [inline Content.Shared/Flight/SharedFlightSystem.cs:110]:
```suggestion
namespace Content.Shared.Flight;

public abstract class SharedFlightSystem : EntitySystem
{
    [Dependency] private readonly SharedActionsSystem _actionsSystem = default!;
    [Dependency] private readonly SharedVirtualItemSystem _virtualItem = default!;
    [Dependency] private readonly StaminaSystem _staminaSystem = default!;
    [Dependency] private readonly SharedHandsSystem _hands = default!;
    [Dependency] private readonly MovementSpeedModifierSystem _movementSpeed = default!;

    public override void Initialize()
    {
        base.Initialize();

        SubscribeLocalEvent<FlightComponent, ComponentStartup>(OnStartup);
        SubscribeLocalEvent<FlightComponent, ComponentShutdown>(OnShutdown);
        SubscribeLocalEvent<FlightComponent, RefreshMovementSpeedModifiersEvent>(OnRefreshMoveSpeed);
    }

    #region Core Functions
    private void OnStartup(EntityUid uid, FlightComponent component, ComponentStartup args)
    {
        _actionsSystem.AddAction(uid, ref component.ToggleActionEntity, component.ToggleAction);
    }

    private void OnShutdown(EntityUid uid, FlightComponent component, ComponentShutdown args)
    {
        _actionsSystem.RemoveAction(uid, component.ToggleActionEntity);
    }

    public void ToggleActive(EntityUid uid, bool active, FlightComponent component)
    {
        component.On = active;
        component.TimeUntilFlap = 0f;
        _actionsSystem.SetToggled(component.ToggleActionEntity, component.On);
        RaiseNetworkEvent(new FlightEvent(GetNetEntity(uid), component.On, component.IsAnimated));
        _staminaSystem.ToggleStaminaDrain(uid, component.StaminaDrainRate, active, false);
        _movementSpeed.RefreshMovementSpeedModifiers(uid);
        UpdateHands(uid, active);
        Dirty(uid, component);
    }

    private void UpdateHands(EntityUid uid, bool flying)
    {
        if (!TryComp<HandsComponent>(uid, out var handsComponent))
            return;

        if (flying)
            BlockHands(uid, handsComponent);
        else
            FreeHands(uid);
    }

    private void BlockHands(EntityUid uid, HandsComponent handsComponent)
    {
        var freeHands = 0;
        foreach (var hand in _hands.EnumerateHands(uid, handsComponent))
        {
            if (hand.HeldEntity == null)
            {
                freeHands++;
                continue;
            }

            // Is this entity removable? (they might have handcuffs on)
            if (HasComp<UnremoveableComponent>(hand.HeldEntity) && hand.HeldEntity != uid)
                continue;

            _hands.DoDrop(uid, hand, true, handsComponent);
            freeHands++;
            if (freeHands == 2)
                break;
        }
        if (_virtualItem.TrySpawnVirtualItemInHand(uid, uid, out var virtItem1))
            EnsureComp<UnremoveableComponent>(virtItem1.Value);

        if (_virtualItem.TrySpawnVirtualItemInHand(uid, uid, out var virtItem2))
            EnsureComp<UnremoveableComponent>(virtItem2.Value);
    }

    private void FreeHands(EntityUid uid)
    {
        _virtualItem.DeleteInHandsMatching(uid, uid);
    }

    private void OnRefreshMoveSpeed(EntityUid uid, FlightComponent component, RefreshMovementSpeedModifiersEvent args)
    {
        Logger.Debug("Refreshing movement speed!");
        if (!component.On)
            return;

        args.ModifySpeed(component.SpeedModifier, component.SpeedModifier);
    }

    #endregion
}
public sealed partial class ToggleFlightEvent : InstantActionEvent { }
```

File: Content.Shared/Gravity/SharedFloatingVisualizerSystem.cs [hunk 29]
```diff
@@ -62,10 +64,23 @@ private void OnGravityChanged(ref GravityChangedEvent args)
...
+    {
+        var uid = GetEntity(args.Uid);
+        if (!TryComp<FloatingVisualsComponent>(uid, out var floating))
+            return;
+        floating.CanFloat = args.IsFlying;
+
+        if (args.IsFlying && args.IsAnimated)
+        {
+            FloatAnimation(uid, floating.Offset, floating.AnimationKey, floating.AnimationTime);
+        }
```
Comment (VMSolidus) [inline Content.Shared/Gravity/SharedFloatingVisualizerSystem.cs:77]:
```suggestion
        if (!args.IsFlying 
            || !args.IsAnimated)
            return;

        FloatAnimation(uid, floating.Offset, floating.AnimationKey, floating.AnimationTime);
```

File: Resources/Locale/en-US/flight/flight_system.ftl [hunk 30]
```diff
@@ -0,0 +1 @@
+no-flight-while-restrained = You can't fly right now.
```
Comment (VMSolidus) [inline Resources/Locale/en-US/flight/flight_system.ftl:1]: Add a localization for no-flight-while-zombified

**Review by VMSolidus — CHANGES_REQUESTED** [review]

**Review by VMSolidus — APPROVED** [review]

**Review by OldDanceJacket — APPROVED** [review]

**Pull request merged**